File: src/main/java/com/example/the_machine/service/ArtifactService.java
package com.example.the_machine.service;

import com.example.the_machine.db.ArtifactEntity;
import com.example.the_machine.db.ArtifactEntity.Kind;
import com.example.the_machine.db.ArtifactEntity.Status;
import com.example.the_machine.domain.event.ArtifactCreatedEvent;
import com.example.the_machine.domain.event.ArtifactScheduledEvent;
import com.example.the_machine.dto.ArtifactMetadataDto;
import com.example.the_machine.exception.ArtifactNotFoundException;
import com.example.the_machine.exception.ArtifactNotReadyException;
import com.example.the_machine.mapper.ArtifactMapper;
import com.example.the_machine.operator.RunContextHolder;
import com.example.the_machine.repo.ArtifactRepository;
import com.example.the_machine.service.events.EventService;
import com.fasterxml.jackson.databind.JsonNode;
import java.io.InputStream;
import java.util.List;
import java.util.UUID;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
@Slf4j
@Transactional(readOnly = true)
public class ArtifactService {

  private final ArtifactRepository artifactRepository;
  private final StorageService storageService;
  private final EventService eventService;
  private final ArtifactMapper artifactMapper;

  /**
   * Get all artifacts for a project
   */
  public List<ArtifactEntity> getProjectArtifacts(UUID projectId) {
    return artifactRepository.findByProjectIdOrderByCreatedAtDesc(projectId);
  }

  /**
   * Get metadata DTOs for all artifacts in a project
   */
  public List<ArtifactMetadataDto> getProjectArtifactMetadata(UUID projectId) {
    return getProjectArtifacts(projectId).stream()
        .map(artifactMapper::toMetadataDto)
        .toList();
  }

  /**
   * Get a specific artifact by ID
   */
  public ArtifactEntity getArtifact(UUID artifactId) {
    return artifactRepository.findById(artifactId)
        .orElseThrow(() -> new ArtifactNotFoundException(
            String.format("Artifact %s not found", artifactId)));
  }

  /**
   * Get metadata DTO for an artifact by ID
   */
  public ArtifactMetadataDto getArtifactMetadata(UUID artifactId) {
    return artifactMapper.toMetadataDto(getArtifact(artifactId));
  }

  /**
   * Download file-based artifact from storage
   */
  public InputStream downloadArtifact(UUID artifactId) {
    final var artifact = getArtifact(artifactId);

    // Validate artifact is file-based and has storage key
    if (artifact.getStorageKey() == null) {
      throw new IllegalStateException(
          String.format("Artifact %s does not have a storage key", artifactId));
    }

    if (artifact.getStatus() != Status.CREATED) {
      throw new ArtifactNotReadyException(
          String.format("Artifact %s is not ready (status: %s)",
              artifactId, artifact.getStatus()));
    }

    log.info("Downloading artifact {} from storage key: {}",
        artifactId, artifact.getStorageKey());

    return storageService.retrieve(artifact.getStorageKey());
  }

  @Transactional
  public UUID schedule(
      Kind kind,
      String title,
      String mime
  ) {
    final var projectId = RunContextHolder.getProjectId();
    final var runId = RunContextHolder.getRunId();
    log.debug("Scheduling artifact for project {}: {}", projectId, title);

    final var artifact = new ArtifactEntity(
        projectId,
        runId,
        kind,
        Status.SCHEDULED,
        title,
        mime,
        null,
        null,
        null,
        null
    );
    artifactRepository.save(artifact);
    eventService.publish(new ArtifactScheduledEvent(projectId, artifact.getId()));

    return artifact.getId();
  }

  @Transactional
  public void updateArtifact(UUID artifactId, JsonNode content, JsonNode metadata) {
    log.debug("Updating artifact {}.\nContent: {} \nMetadata: {}", artifactId, content, metadata);
    final var artifact = getArtifact(artifactId);
    artifact.setContent(content);
    artifact.setMetadata(metadata);
    artifact.setStatus(Status.CREATED);
    artifactRepository.save(artifact);

    eventService.publish(new ArtifactCreatedEvent(artifact.getProjectId(), artifact.getId()));
  }
}

File: src/main/kotlin/com/example/the_machine/service/ArtifactService.kt
package com.example.the_machine.service

import com.example.the_machine.common.IdGenerator
import com.example.the_machine.domain.ArtifactEntity
import com.example.the_machine.repo.ArtifactRepository
import com.fasterxml.jackson.databind.JsonNode
import mu.KotlinLogging
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.Instant
import java.util.*

private val log = KotlinLogging.logger {}

/**
 * Service for handling artifact creation and management using direct type-safe methods.
 */
@Service
class ArtifactService(
  private val artifactRepository: ArtifactRepository,
  private val idGenerator: IdGenerator
) {

  // Direct artifact creation methods

  /**
   * Creates an artifact directly with metadata - no JSON parsing needed.
   *
   * @param threadId the thread ID
   * @param runId the run ID
   * @param kind the artifact kind
   * @param stage the workflow stage
   * @param title the artifact title
   * @param mime the MIME type (optional)
   * @return the created artifact entity
   */
  @Transactional
  fun createArtifact(
    threadId: UUID,
    runId: UUID,
    kind: ArtifactEntity.Kind,
    stage: ArtifactEntity.Stage,
    title: String,
    mime: String?
  ): ArtifactEntity {
    log.debug { "Creating artifact: kind=$kind, stage=$stage, title=$title" }

    val artifact = ArtifactEntity(
      id = idGenerator.newId(),
      threadId = threadId,
      runId = runId,
      kind = kind,
      stage = stage,
      status = ArtifactEntity.Status.PENDING,
      title = title,
      mime = mime,
      createdAt = Instant.now()
    )

    val saved = artifactRepository.save(artifact)
    log.info { "Artifact created: ${saved.id} for thread: $threadId" }
    return saved
  }

  /**
   * Updates artifact with structured content (JSON).
   *
   * @param artifactId the artifact ID
   * @param content the content to set
   */
  @Transactional
  fun setContent(artifactId: UUID, content: JsonNode) {
    val artifact = artifactRepository.findById(artifactId)
      .orElseThrow { IllegalArgumentException("Artifact not found: $artifactId") }

    val updatedArtifact = artifact.copy(
      content = content,
      status = ArtifactEntity.Status.DONE
    )
    artifactRepository.save(updatedArtifact)

    log.debug { "Content set for artifact: $artifactId" }
  }

  /**
   * Updates artifact with file storage information.
   *
   * @param artifactId the artifact ID
   * @param storageKey the storage location
   * @param mime the MIME type
   */
  @Transactional
  fun setStorageKey(artifactId: UUID, storageKey: String, mime: String) {
    val artifact = artifactRepository.findById(artifactId)
      .orElseThrow { IllegalArgumentException("Artifact not found: $artifactId") }

    val updatedArtifact = artifact.copy(
      storageKey = storageKey,
      mime = mime,
      status = ArtifactEntity.Status.DONE
    )
    artifactRepository.save(updatedArtifact)

    log.debug { "Storage key set for artifact: $artifactId -> $storageKey" }
  }

  /**
   * Updates artifact metadata.
   *
   * @param artifactId the artifact ID
   * @param metadata the metadata to set
   */
  @Transactional
  fun setMetadata(artifactId: UUID, metadata: JsonNode) {
    val artifact = artifactRepository.findById(artifactId)
      .orElseThrow { IllegalArgumentException("Artifact not found: $artifactId") }

    val updatedArtifact = artifact.copy(metadata = metadata)
    artifactRepository.save(updatedArtifact)

    log.debug { "Metadata set for artifact: $artifactId" }
  }
}